bin/ch-completion.bash

```diff
@@ -0,0 +1,388 @@
+    echo "does nothing"
+}
+
+## Helper functions ##
+
+# Use this function to enable or disable completion after sourcing
+# this file. Useful if there are issues.
+# FIXME: Add other executables as their completion functions are
+#        implemented.
+ch-completion () {
+    if [[ "$1" == "disable" ]]; then
+        complete -r ch-image
+    elif [[ "$1" == "enable" ]]; then
+        complete -F _ch_image_completion ch-image
+    fi
+}
+
+# Figure out which storage directory to use (including cli-specified
+# storage).
+# FIXME: Can probably cook up a sed pattern that'll remove the need
+#        for the “if” statement.
+_ch_find_storage () {
+    if [[ -n "$(grep -Eo "(\-\-storage|[^\-]\-s)" <<< "$@")" ]]; then
+        # This sed only works as desired if “--storage” or “-s” are present
+        # in the command line...
+        sed -E 's/(.*)(--storage|[^-]-s)\ *([^ ]*)(.*$)/\3/g' <<< "$@"
+    elif [[ -n "$CH_IMAGE_STORAGE" ]]; then
+        echo "$CH_IMAGE_STORAGE"
+    else
+        echo "/var/tmp/$USER.ch/"
+    fi
+}
+
+# Horrible, disgusting function to find an image or image ref in
+# the ch-run command line.
+#
+# NOT FINISHED, DON'T USE!!!
+_ch_run_image_finder () {
+    # Takes array of words. Tries to find an image in there.
+    #local -n words=$1
+    local wrds=("$@")
+    local ct=1
+
+    #echo "len: ${#wrds[@]}"
+
+    while (($ct < ${#wrds[@]})); do
+        if [[ ( -f "${wrds[$ct]}" && ( "${wrds[$ct]}" == *".sqfs" || "${wrds[$ct]}" == *".tar."* || "${wrds[$ct]}" == *".tgz" ) ) || ( -d "${wrds[$ct]}" && "${wrds[$ct-1]}" != "--mount" &&
+                "${wrds[$ct-1]}" != "-m" && "${wrds[$ct-1]}" != "--bind" &&
+                "${wrds[$ct-1]}" != "-b" ) ]]; then
+            echo "${wrds[$ct]}"
+        fi
+        # FIXME: Check for image refs
+        ((ct++))
+    done
+}
+
+# Kludge up a way to look through array of words and determine the
+# subcommand. Note that the double for loop doesn't take that much
+# time, since the Charliecloud command line is relatively short.
+#
+# Usage:
+#   _ch_subcommand_get [subcommands] [words]
+#
+# Example:
+#   >> _ch_subcommand_get "build build-cache ... undelete" \
+#                         "ch-image --foo build ..."
+#      build
+_ch_subcommand_get () {
+    local cmd subcmd
+    local cmds="$1"
+    shift 1
+    for word in "$@"
+    do
+        echo "word: $word" >> /tmp/ch-completion.log
+        for cmd in $cmds
+        do
+            if [[ "$word" == "$cmd" ]]; then
+                subcmd="$cmd"
+            fi
+        done
+    done
+    echo "$subcmd"
+}
+
+# Code that I shamelessly stole from StackOverflow
+#   (https://stackoverflow.com/a/40227233)
+# Returns filenames and directories, appending a slash to directory names.
+# This function takes option “-X,” a string of space-separated glob patterns
+# to be excluded from file completion using the compgen option of the same
+# name (see https://devdocs.io/bash/programmable-completion-builtins#index-compgen)
```

Comment:
Use imperative case.

As for “shamelessly stole”, I’d just put the link at the end of the comment.